File: advent-of-code/2025/day-8-playground/normal/n-d08-12.kt
import kotlin.math.sqrt
import kotlin.math.pow

class UF(n: Int) {
    val p = IntArray(n) { it }
    val s = IntArray(n) { 1 }
    fun find(x: Int): Int { if (p[x] != x) p[x] = find(p[x]); return p[x] }
    fun unite(x: Int, y: Int): Boolean {
        var px = find(x); var py = find(y)
        if (px == py) return false
        if (s[px] < s[py]) { val t = px; px = py; py = t }
        p[py] = px; s[px] += s[py]; return true
    }
}

fun main() {
    val pts = generateSequence(:: readLine).map { l -> l.split(",").map { it.toInt() } }. toList()
    val n = pts.size
    val dists = mutableListOf<Triple<Double, Int, Int>>()
    for (i in 0 until n) for (j in i + 1 until n) {
        val d = sqrt((0.. 2).sumOf { (pts[i][it] - pts[j][it]).toDouble().pow(2) })
        dists.add(Triple(d, i, j))
    }
    dists.sortBy { it.first }
    
    val uf1 = UF(n)
    dists.take(1000).forEach { (_, i, j) -> uf1.unite(i, j) }
    val sizes = (0 until n).filter { uf1.find(it) == it }.map { uf1.s[it] }. sortedDescending()
    println(sizes[0]. toLong() * sizes[1] * sizes[2])
    
    val uf2 = UF(n)
    for ((_, i, j) in dists) {
        if (uf2.unite(i, j) && (0 until n).map { uf2.find(it) }.toSet().size == 1) {
            println(pts[i][0]. toLong() * pts[j][0])
            break
        }
    }
}

File: advent-of-code/2025/day-8-playground/normal/n-d08-12.rs
use std::io::{self, BufRead};
use std::collections::HashSet;

struct UF { p: Vec<usize>, s: Vec<usize> }
impl UF {
    fn new(n: usize) -> Self { UF { p: (0..n).collect(), s: vec![1; n] } }
    fn find(&mut self, x: usize) -> usize {
        if self.p[x] != x { self.p[x] = self.find(self.p[x]); }
        self.p[x]
    }
    fn unite(&mut self, x: usize, y: usize) -> bool {
        let (mut px, mut py) = (self.find(x), self.find(y));
        if px == py { return false; }
        if self.s[px] < self.s[py] { std::mem::swap(&mut px, &mut py); }
        self.p[py] = px; self.s[px] += self.s[py]; true
    }
}

fn main() {
    let pts: Vec<Vec<i64>> = io::stdin().lock().lines().flatten()
        .map(|l| l.split(',').map(|x| x.parse().unwrap()).collect()).collect();
    let n = pts.len();
    let mut dists: Vec<(f64, usize, usize)> = vec![];
    for i in 0..n { for j in i+1..n {
        let d = ((0.. 3).map(|k| (pts[i][k] - pts[j][k]).pow(2) as f64).sum::<f64>()).sqrt();
        dists.push((d, i, j));
    }}
    dists.sort_by(|a, b| a. 0.partial_cmp(&b.0).unwrap());
    
    let mut uf1 = UF::new(n);
    for &(_, i, j) in dists.iter().take(1000) { uf1.unite(i, j); }
    let mut sizes:  Vec<usize> = (0..n).filter(|&i| uf1.find(i) == i).map(|i| uf1.s[i]).collect();
    sizes.sort_by(|a, b| b.cmp(a));
    println!("{}", sizes[0] as i64 * sizes[1] as i64 * sizes[2] as i64);
    
    let mut uf2 = UF::new(n);
    for &(_, i, j) in &dists {
        if uf2.unite(i, j) {
            let roots: HashSet<usize> = (0..n).map(|k| uf2.find(k)).collect();
            if roots.len() == 1 { println!("{}", pts[i][0] * pts[j][0]); break; }
        }
    }
}
